packages/web/src/common/store/cache/collections/createAlbumSaga.ts

```diff
@@ -0,0 +1,257 @@
+  yield* takeLatest(cacheCollectionsActions.CREATE_PLAYLIST, createAlbumWorker)
+}
+
+function* createAlbumWorker(
+  action: ReturnType<
+    | typeof cacheCollectionsActions.createAlbum
+    | typeof cacheCollectionsActions.createPlaylist
+  >
+) {
+  // Return early if this is not an album
+  if (!action.isAlbum) return
+
+  yield* waitForWrite()
+  const userId = yield* call(ensureLoggedIn)
+  const { initTrackId, formFields, source, noticeType } = action
+  const collection = newCollectionMetadata({ ...formFields, is_album: true })
+  const sdk = yield* getSDK()
+  const collectionId = yield* call([
+    sdk.playlists,
+    sdk.playlists.generatePlaylistId
+  ])
+  if (!collectionId) return
+
+  const initTrack = yield* select(getTrack, { id: initTrackId })
+
+  if (initTrack) {
+    collection.cover_art_sizes = initTrack.cover_art_sizes
+  }
+
+  yield* call(optimisticallySaveAlbum, collectionId, collection, initTrack)
+  yield* put(
+    cacheCollectionsActions.createPlaylistRequested(
+      collectionId,
+      noticeType,
+      true
+    )
+  )
+  yield* call(
+    createAndConfirmAlbum,
+    collectionId,
+    userId,
+    collection,
+    initTrack,
+    source
+  )
+}
+
+function* optimisticallySaveAlbum(
+  albumId: ID,
+  formFields: Partial<CollectionMetadata>,
+  initTrack: Nullable<Track>
+) {
+  const accountUser = yield* select(getAccountUser)
+  if (!accountUser) return
+  const { user_id, handle, _collectionIds = [] } = accountUser
+  const album: Partial<Collection> = {
+    playlist_id: albumId,
+    ...formFields
+  }
+
+  const initTrackOwner = yield* select(getUser, { id: initTrack?.owner_id })
+
+  album.playlist_owner_id = user_id
+  album.is_private = true
+  album.playlist_contents = {
+    track_ids: initTrack
+      ? [
+          {
+            time: Math.round(Date.now() / 1000),
+            track: initTrack.track_id
+          }
+        ]
+      : []
+  }
+  album.tracks = initTrack
+    ? [
+        {
+          ...initTrack,
+          user: initTrackOwner!
+        }
+      ]
+    : []
+  album.track_count = initTrack ? 1 : 0
+  album.permalink = collectionPage(
+    handle,
+    album.playlist_name,
+    albumId,
+    undefined,
+    true
+  )
+
+  yield* put(
+    cacheActions.add(
+      Kind.COLLECTIONS,
+      [{ id: albumId, metadata: album }],
+      true,
+      false
+    )
+  )
+
+  yield* put(
+    cacheActions.update(Kind.USERS, [
+      {
+        id: user_id,
+        metadata: { _collectionIds: _collectionIds.concat(albumId) }
+      }
+    ])
+  )
+}
+
+function* createAndConfirmAlbum(
+  albumId: ID,
+  userId: ID,
+  formFields: EditCollectionValues,
+  initTrack: Nullable<Track>,
+  source: string
+) {
+  const sdk = yield* getSDK()
+
+  const event = make(Name.PLAYLIST_START_CREATE, {
+    source,
+    artworkSource: formFields.artwork
+      ? formFields.artwork.source
+      : formFields.cover_art_sizes
+  })
+  yield* put(event)
+
+  function* confirmAlbum() {
```

Comment:
I like that you kept parity with the playlist saga and don't think you should change it, but just sharing that SDK does the write confirmation internally! Something to keep in mind for tan-query work

Reply:
yeah great reminder, we have used this to great effect in tan-query